AsyncDisplayKit/AsyncDisplayKit+Debug.m

```diff
@@ -191,3 +203,551 @@ - (void)drawEdgeIfClippedWithEdges:(UIRectEdge)rectEdge color:(UIColor *)color b
+    fetchDataTuningParameters:(ASRangeTuningParameters)fetchDataTuningParameters
+               interfaceState:(ASInterfaceState)interfaceState;
+
+@end
+
+@interface _ASRangeDebugBarView : UIView
+
+@property (nonatomic, weak) ASRangeController *rangeController;
+@property (nonatomic, assign) BOOL destroyOnLayout;
+@property (nonatomic, strong) NSString *debugString;
+
+- (instancetype)initWithRangeController:(ASRangeController *)rangeController;
+
+- (void)updateWithVisibleRatio:(CGFloat)visibleRatio
+                  displayRatio:(CGFloat)displayRatio
+           leadingDisplayRatio:(CGFloat)leadingDisplayRatio
+                fetchDataRatio:(CGFloat)fetchDataRatio
+         leadingFetchDataRatio:(CGFloat)leadingFetchDataRatio
+                     direction:(ASScrollDirection)direction;
+
+@end
+
+static BOOL __shouldShowRangeDebugOverlay = NO;
+
+@implementation ASRangeController (Debugging)
+
++ (void)setShouldShowRangeDebugOverlay:(BOOL)show
+{
+  __shouldShowRangeDebugOverlay = show;
+}
+
++ (BOOL)shouldShowRangeDebugOverlay
+{
+  return __shouldShowRangeDebugOverlay;
+}
+
++ (void)layoutDebugOverlayIfNeeded
+{
+  [[_ASRangeDebugOverlayView sharedInstance] setNeedsLayout];
+}
+
+- (void)addRangeControllerToRangeDebugOverlay
+{
+  [[_ASRangeDebugOverlayView sharedInstance] addRangeController:self];
+}
+
+- (void)updateRangeController:(ASRangeController *)controller
+     withScrollableDirections:(ASScrollDirection)scrollableDirections
+              scrollDirection:(ASScrollDirection)direction
+                    rangeMode:(ASLayoutRangeMode)mode
+      displayTuningParameters:(ASRangeTuningParameters)displayTuningParameters
+    fetchDataTuningParameters:(ASRangeTuningParameters)fetchDataTuningParameters
+               interfaceState:(ASInterfaceState)interfaceState
+{
+  [[_ASRangeDebugOverlayView sharedInstance] updateRangeController:controller
+                                          withScrollableDirections:scrollableDirections
+                                                   scrollDirection:direction
+                                                         rangeMode:mode
+                                           displayTuningParameters:displayTuningParameters
+                                         fetchDataTuningParameters:fetchDataTuningParameters
+                                                    interfaceState:interfaceState];
+}
+
+@end
+
+
+#pragma mark _ASRangeDebugOverlayView
+
+@interface _ASRangeDebugOverlayView () <UIGestureRecognizerDelegate>
+@end
+
+@implementation _ASRangeDebugOverlayView
+{
+  NSMutableArray *_rangeControllerViews;
+  NSInteger      _newControllerCount;
+  NSInteger      _removeControllerCount;
+  BOOL           _animating;
+}
+
++ (UIWindow *)keyWindow
+{
+  // hack to work around app extensions not having UIApplication...not sure of a better way to do this?
+  return [[NSClassFromString(@"UIApplication") sharedApplication] keyWindow];
```

Comment:
I think this is fine for now. If we have any issues, we could add some API like `setShouldShowRangeDebugOverlay:inWindow:` or something but this is fine.